### PR TITLE
Enable page refresh at every step

### PR DESCRIPTION
**Pull request by sisou**
With these changes, and the according changes in the @nimiq/rpc and @nimiq/keyguard-client libs (already included here as package dependencies), it becomes possible for the user to refresh the page at every step throughout a request flow, without loosing the request or getting stuck.

### CODE REVIEW

File: client/RequestBehavior.ts [hunk 1]
```diff
@@ -97,11 +97,51 @@ export class PopupRequestBehavior extends RequestBehavior {
...
+
+        try {
+            const result = await client.listenFor(command);
+            client.close();
+            popup.close();
+            return result;
+        } catch (e) {
+            client.close();
+            popup.close();
+            throw e;
```
[the remark this answers is not in the document]
Comment (sisou) [reply client/RequestBehavior.ts]: Does the `finally` block still run even when the function `return`s in the `try` block?

[the remark this answers is not in the document]
Comment (sisou) [reply client/RequestBehavior.ts]: I just looked it up on MDN and the answer should be yes. In that case, yes, let's do that.

[the remark this answers is not in the document]
Comment (danimoh) [reply client/RequestBehavior.ts]: Yes, `finally` is pretty cool

File: src/lib/RpcApi.ts [hunk 2]
```diff
@@ -62,6 +62,16 @@ export default class RpcApi {
         return client;
     }
 
+    public routerPush(routeName: string) {
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/RpcApi.ts]: They need to be accessible from every component, thus the Vue prototype property `.$rpc` lends itself perfectly to thus oppertunity, as otherwise we would have to extend the native `vue-router`, which would also be available in each component as `.$router`. These `routerPush` and `routerReplace` functions however are wrappers for the `vue-router`'s native `push` and `replace` methods. Thus it's a bit hard to come up with better names on the router instance itself, but it would still be prototype-extension, which is generally considered bad practice.

[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/RpcApi.ts]: And it also actually has to do with the RpcApi itself, because it's used for preserving the rpc parameters of the request and responses. Thus it's not _that_ bad a place for the methods.

File: client/RequestBehavior.ts [hunk 3]
```diff
@@ -77,14 +77,12 @@ export class PopupRequestBehavior extends RequestBehavior {
         await client.init();
 
         try {
-            const result = await client.call(command, ...args);
-            client.close();
-            popup.close();
-            return result;
+            return await client.callAndPersist(command, ...args);
```
Comment (sisou) [inline client/RequestBehavior.ts]: Note for reviewers: this is basically where the magic is happening. The new `callAndPersist()` method makes sure that the receiving AccountsManager RpcServer writes the request into the URL after validating it, thus enabling a page refresh in the AccountsManager (after a page refresh, the request would be read from the URL as if it were a redirect request, but a special `returnUrl` parameter signals that the request should be replied to via postMessage).

File: src/main.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ self.NIMIQ_IQONS_SVG_PATH = '/img/iqons.min.cc877caa.svg';
 
 const rpcApi = new RpcApi(store, staticStore, router);
 Vue.prototype.$rpc = rpcApi;
+// rpcApi ist started in App.vue->created()
```
Comment (nibhar) [inline src/main.ts]:
```suggestion
// rpcApi is started in App.vue->created()
```

File: client/package.json [hunk 5]
```diff
@@ -8,7 +8,7 @@
   "license": "Apache-2.0",
   "private": false,
   "dependencies": {
-    "@nimiq/rpc": "^0.0.11"
+    "@nimiq/rpc": "^0.1.0-beta.5"
```
[the remark this answers is not in the document]
Comment (sisou) [reply client/package.json]: RPC release pending here: https://github.com/nimiq/rpc/pull/8

File: package.json [hunk 6]
```diff
@@ -5,16 +5,16 @@
...
         "test:unit": "vue-cli-service test:unit",
         "test": "yarn test:unit",
         "pr": "yarn lint && yarn build:ci && yarn test && cd client && yarn lint && yarn build"
     },
     "dependencies": {
-        "@nimiq/keyguard-client": "^0.1.2",
+        "@nimiq/keyguard-client": "^0.1.5",
         "@nimiq/network-client": "^0.1.0",
-        "@nimiq/rpc": "^0.0.11",
+        "@nimiq/rpc": "^0.1.0-beta.4",
```
[the remark this answers is not in the document]
Comment (sisou) [reply package.json]: RPC release pending here: https://github.com/nimiq/rpc/pull/8

File: src/components/Network.vue [hunk 7]
```diff
@@ -59,7 +59,6 @@ class Network extends Vue {
                     new Nimiq.PublicKey(keyguardResult.publicKey),
                     new Nimiq.Signature(keyguardResult.signature),
                 ).serialize(),
-                keyguardRequest.networkId,
```
Comment (Bettelstab) [inline src/components/Network.vue]: I don't think this is related to page refresh, is it?

File: src/lib/RpcApi.ts [hunk 8]
```diff
@@ -114,11 +130,25 @@ export default class RpcApi {
...
+                this.routerReplace(request);
             });
         }
     }
 
+    private _parseUrlParams(query: string) {
+        const params: {[key: string]: string} = {};
+        if (!query) return params;
+        const keyValues = query.substr(1).replace(/\+/g, ' ').split('&')
+            .map((keyValueString) => keyValueString.split('='));
```
Comment (Bettelstab) [inline src/lib/RpcApi.ts]: You could use URLSearchParams here.
```suggestion\
           const keyValues = new URLSearchParams(query);
```

File: src/views/CheckoutOverview.vue [hunk 9]
```diff
@@ -75,7 +76,6 @@ export default class CheckoutOverview extends Vue {
             validityStartHeight,
             data: this.request.data,
             flags: this.request.flags,
-            networkId: this.request.networkId,
```
Comment (Bettelstab) [inline src/views/CheckoutOverview.vue]: unrelated

File: src/views/SignTransaction.vue [hunk 10]
```diff
@@ -43,7 +43,6 @@ export default class SignTransaction extends Vue {
             validityStartHeight: this.request.validityStartHeight,
             data: this.request.data,
             flags: this.request.flags,
-            networkId: this.request.networkId,
```
Comment (Bettelstab) [inline src/views/SignTransaction.vue]: unrelated

File: src/components/Network.vue (hunk 7 above)
Comment (sisou) [reply src/components/Network.vue]: It is, because I am updating the KeyguardClient dependency, to have the RPC features required, but the API there changed and does not include the networkId anymore. Thus, to use the new KeyguardClient, this needs to be removed (will otherwise throw a typescript error).

File: src/views/CheckoutOverview.vue (hunk 9 above)
Comment (sisou) [reply src/views/CheckoutOverview.vue]: It is related to updating the KeyguardClient dependency, to have the RPC features required. The API there changed and does not include the networkId anymore. Thus, to use the new KeyguardClient, this needs to be removed (will otherwise throw a typescript error).

File: src/views/SignTransaction.vue (hunk 10 above)
Comment (sisou) [reply src/views/SignTransaction.vue]: It is related to updating the KeyguardClient dependency, to have the RPC features required. The API there changed and does not include the networkId anymore. Thus, to use the new KeyguardClient, this needs to be removed (will otherwise throw a typescript error).

File: src/lib/RpcApi.ts (hunk 8 above)
Comment (sisou) [reply src/lib/RpcApi.ts]: `URLSearchParams` decodes the params differently from what vue-router does (https://github.com/vuejs/vue-router/blob/dev/src/util/query.js#L65), which is exactly the problem I created this function for. Otherwise it would have been easy :laughing:.

Comment (Bettelstab) [reply src/lib/RpcApi.ts]: I see :D
